File: tests/clean/test_hogbom_clean.py
# See the LICENSE file at the top-level directory of this distribution.

"""Integration test for Hogbom CLEAN using DFT and gridder functions."""

try:
    import cupy
except ImportError:
    cupy = None

import numpy as np
import scipy.signal as sig

from ska_sdp_func.clean import hogbom_clean
from ska_sdp_func.grid_data import GridderUvwEsFft
from ska_sdp_func.visibility import dft_point_v01


def create_test_data(dirty_size, psf_size):
    """create a test image and corresponding PSF using PFL functions.
    This image and PSF can then be used to test the Hogbom CLEAN
    against known data.
    Creates a simulated UVW coverage of randomly sized baselines,
    random lmn coordinates for sources are generated and processed
    by DFT and gridder functions in to a dirty image."""
    # Initialise settings
    num_components = 10
    num_pols = 1
    num_channels = 1
    num_baselines = 2000
    num_times = 1
    channel_start_hz = 100e6
    channel_step_hz = 100e3
    np.random.seed(12)  # seed for generating data

    # initialise empty output array for visibilities
    vis = np.zeros(
        [num_times, num_baselines, num_channels, num_pols],
        dtype=np.complex128,
    )

    vis_psf = np.zeros(
        [num_times, num_baselines, num_channels, num_pols],
        dtype=np.complex128,
    )

    # initialise empty arrays for intermediate data
    fluxes = np.zeros([num_components, num_channels, num_pols]) + 0j

    fluxes_psf = np.ones([1, num_channels, num_pols]) + 0j
    directions = np.zeros([num_components, 3])
    directions_psf = np.zeros([1, 3])
    uvw = np.zeros([num_times, num_baselines, 3])

    # UVW coverage as a random filled circle of given length, centered on (0,0)
    y0 = 0
    x0 = 0
    r = 3000

    # random filled circle representing baselines
    for i in range(num_baselines):

        # random angle
        theta = 2 * np.pi * np.random.uniform()

        # random radius
        radius = r * np.random.uniform(0, 1)

        x = x0 + radius * np.cos(theta)
        y = y0 + radius * np.sin(theta)

        uvw[0, i, :] = x, y, 0

    # Create random fluxes between 0 + 0j and 10 + 0j for sources
    for i in range(num_components):
        fluxes[i, 0, 0] = np.random.uniform(1, 10) + 0j

    # create random lmn co-ordinates between (-0.015,-0.015) and (0.015,0.015)
    for i in range(num_components):
        directions[i, :] = (
            np.random.uniform(-0.015, 0.015),
            np.random.uniform(-0.015, 0.015),
            0,
        )

        directions[i, 2] = np.sqrt(
            1 - (directions[i, 0]) ** 2 - (directions[i, 1] ** 2)
        )

    # process in to visibilities
    print("Creating visibilities using dft_point_v01 from ska-sdp-func...")
    # dirty image
    dft_point_v01(
        directions, fluxes, uvw, channel_start_hz, channel_step_hz, vis
    )

    # psf
    dft_point_v01(
        directions_psf,
        fluxes_psf,
        uvw,
        channel_start_hz,
        channel_step_hz,
        vis_psf,
    )

    # grid the generated visibilities
    # initialise settings for gridder
    nxydirty = dirty_size
    nxydirty_psf = psf_size
    fov = 2  # degrees
    pixel_size_rad = fov * np.pi / 180 / nxydirty
    pixel_size_rad_psf = fov * np.pi / 180 / nxydirty_psf
    freqs = channel_start_hz + np.arange(num_channels) * channel_step_hz
    do_w_stacking = False
    epsilon = 1e-5
    num_vis = vis.shape[1]

    # reshape vis output from DFT to fit input of gridder
    vis = vis[0, :, :]
    vis_psf = vis_psf[0, :, :]

    # reshape UVW to fit input of gridder
    uvw = uvw[0, :, :]

    # Run gridder test on GPU, using cupy arrays.
    if cupy:
        freqs_gpu = cupy.asarray(freqs)
        uvw_gpu = cupy.asarray(uvw)
        dirty_image_gpu = cupy.zeros([nxydirty, nxydirty], np.float64)
        psf_gpu = cupy.zeros([nxydirty_psf, nxydirty_psf], np.float64)
        vis_gpu = cupy.array(vis, np.complex128, order="C")
        vis_psf_gpu = cupy.array(vis_psf, np.complex128, order="C")
        weight_gpu = cupy.ones([num_vis, num_channels])

        # Create gridder for dirty image
        gridder = GridderUvwEsFft(
            uvw_gpu,
            freqs_gpu,
            vis_gpu,
            weight_gpu,
            dirty_image_gpu,
            pixel_size_rad,
            pixel_size_rad,
            epsilon,
            do_w_stacking,
        )

        # Run gridder
        print(
            "Creating Dirty image using GridderUvwEsFft from ska-sdp-func..."
        )
        gridder.grid_uvw_es_fft(
            uvw_gpu, freqs_gpu, vis_gpu, weight_gpu, dirty_image_gpu
        )

        # Create gridder for PSF
        gridder = GridderUvwEsFft(
            uvw_gpu,
            freqs_gpu,
            vis_psf_gpu,
            weight_gpu,
            psf_gpu,
            pixel_size_rad_psf,
            pixel_size_rad_psf,
            epsilon,
            do_w_stacking,
        )

        # Run gridder
        print("Creating PSF using GridderUvwEsFft from ska-sdp-func...")
        gridder.grid_uvw_es_fft(
            uvw_gpu, freqs_gpu, vis_psf_gpu, weight_gpu, psf_gpu
        )

        dirty_image = cupy.asnumpy(dirty_image_gpu)
        psf = cupy.asnumpy(psf_gpu)

    # normalise fluxes by number of baselines
    psf = psf / num_baselines
    dirty_image = dirty_image / num_baselines

    return dirty_image, psf


def create_cbeam(coeffs, size):
    """create clean beam"""

    center = size // 2

    cbeam = np.zeros([size, size])

    A = 1
    x0 = center
    y0 = center
    sigma_X = coeffs[0]
    sigma_Y = coeffs[1]
    theta = (np.pi / 180) * coeffs[2]

    X = np.arange(0, size, 1)
    Y = np.arange(0, size, 1)

    a = np.cos(theta) ** 2 / (2 * sigma_X ** 2) + np.sin(theta) ** 2 / (
        2 * sigma_Y ** 2
    )
    b = np.sin(2 * theta) / (4 * sigma_X ** 2) - np.sin(2 * theta) / (
        4 * sigma_Y ** 2
    )
    c = np.sin(theta) ** 2 / (2 * sigma_X ** 2) + np.cos(theta) ** 2 / (
        2 * sigma_Y ** 2
    )

    for x in X:
        for y in Y:
            cbeam[x, y] = A * np.exp(
                -(
                    a * (x - x0) ** 2
                    + 2 * b * (x - x0) * (y - y0)
                    + c * (y - y0) ** 2
                )
            )

    return cbeam


def reference_hogbom_clean(
    dirty_img, psf, cbeam_details, loop_gain, threshold, cycle_limit
):
    """Reference python implementation of Hogbom CLEAN to
    test against c version"""
    # calculate useful shapes and sizes
    dirty_size = dirty_img.shape[0]

    # set up so loop variables
    cur_cycle = 0
    stop = False

    # Create intermediate data arrays
    maximum = np.zeros((cycle_limit, 3))
    clean_comp = np.zeros(dirty_img.shape)
    residual = np.copy(dirty_img)

    # create CLEAN beam
    cbeam = create_cbeam(cbeam_details, dirty_size)

    # begin CLEAN while thereshold and number of iterations not exceded
    while cur_cycle < cycle_limit and stop is False:

        # Find index of the maximum value in residual
        max_idx_flat = residual.argmax()
        max_idx = np.unravel_index(max_idx_flat, residual.shape)

        # check maximum value against threshold
        if residual[max_idx] < threshold:
            # Set stop flag if threshold reached
            stop = True
            break

        # Save position and peak value
        maximum[cur_cycle] = residual[max_idx], max_idx[0], max_idx[1]

        # Add fraction of maximum to clean components list
        clean_comp[max_idx[0], max_idx[1]] += loop_gain * maximum[cur_cycle, 0]

        # Shift the center of the PSF to the position of the maxmium
        shifted_psf = psf[
            (dirty_size - max_idx[0]) : (dirty_size - max_idx[0] + dirty_size),
            (dirty_size - max_idx[1]) : (dirty_size - max_idx[1] + dirty_size),
        ]

        # Subtract scaled version of PSF from residual (loop gain x maximum)
        residual = np.subtract(
            residual, (loop_gain * residual[max_idx] * shifted_psf)
        )

        cur_cycle += 1

    # Convolve clean beam with clean components
    skymodel = sig.convolve(clean_comp, cbeam, mode="same")

    # Add remaining residual
    skymodel = np.add(skymodel, residual)

    return skymodel, residual, clean_comp


def test_hogbom_clean():
    """Test the Hogbom CLEAN function python version against c"""

    # initalise settings
    dirty_size = 256
    psf_size = 512
    cbeam_details = np.array([2.0, 2.0, 1.0, 128.0], dtype=np.float64)
    loop_gain = 0.1
    threshold = 0.001
    cycle_limit = 10000

    # create empty array for result
    skymodel = np.zeros((dirty_size, dirty_size))
    clean_model = np.zeros((dirty_size, dirty_size))
    residual = np.zeros((dirty_size, dirty_size))

    # create test data
    print("Creating test data on CPU from ska-sdp-func...")
    dirty_img, psf = create_test_data(dirty_size, psf_size)

    print("Creating reference data on CPU from ska-sdp-func...")
    (
        skymodel_reference,
        residual_reference,
        clean_model_reference,
    ) = reference_hogbom_clean(
        dirty_img, psf, cbeam_details, loop_gain, threshold, cycle_limit
    )

    print("Reference created")

    print("Testing Hogbom CLEAN on CPU from ska-sdp-func...")

    hogbom_clean(
        dirty_img,
        psf,
        cbeam_details,
        loop_gain,
        threshold,
        cycle_limit,
        clean_model,
        residual,
        skymodel,
    )

    np.testing.assert_array_almost_equal(
        clean_model, clean_model_reference, decimal=6
    )
    np.testing.assert_array_almost_equal(
        residual, residual_reference, decimal=6
    )
    np.testing.assert_array_almost_equal(
        skymodel, skymodel_reference, decimal=6
    )

    print("Hogbom CLEAN double precision on CPU: Test passed")

    dirty_img_float = dirty_img.astype(np.float32)
    psf_float = psf.astype(np.float32)
    skymodel_float = skymodel.astype(np.float32)
    clean_model_float = clean_model.astype(np.float32)
    residual_float = residual.astype(np.float32)
    cbeam_details_float = cbeam_details.astype(np.float32)

    hogbom_clean(
        dirty_img_float,
        psf_float,
        cbeam_details_float,
        loop_gain,
        threshold,
        cycle_limit,
        clean_model_float,
        residual_float,
        skymodel_float,
    )

    np.testing.assert_array_almost_equal(
        clean_model_float, clean_model_reference, decimal=4
    )
    np.testing.assert_array_almost_equal(
        residual_float, residual_reference, decimal=4
    )
    np.testing.assert_array_almost_equal(
        skymodel_float, skymodel_reference, decimal=4
    )

    print("Hogbom CLEAN float precision on CPU: Test passed")

    if cupy:
        dirty_img_gpu = cupy.asarray(dirty_img)
        psf_gpu = cupy.asarray(psf)
        # cbeam_details_gpu = cupy.asarray(cbeam_details)
        clean_model_gpu = cupy.zeros_like(dirty_img_gpu)
        residual_gpu = cupy.zeros_like(dirty_img_gpu)
        skymodel_gpu = cupy.zeros_like(dirty_img_gpu)

        print("Testing Hogbom CLEAN on GPU from ska-sdp-func...")

        hogbom_clean(
            dirty_img_gpu,
            psf_gpu,
            cbeam_details,
            loop_gain,
            threshold,
            cycle_limit,
            clean_model_gpu,
            residual_gpu,
            skymodel_gpu,
        )

        clean_model_check = cupy.asnumpy(clean_model_gpu)
        residual_check = cupy.asnumpy(residual_gpu)
        skymodel_check = cupy.asnumpy(skymodel_gpu)

        np.testing.assert_array_almost_equal(
            clean_model_check, clean_model_reference, decimal=6
        )
        np.testing.assert_array_almost_equal(
            residual_check, residual_reference, decimal=6
        )
        np.testing.assert_array_almost_equal(
            skymodel_check, skymodel_reference, decimal=6
        )

        print("Hogbom CLEAN double precision on GPU: Test passed")

        dirty_img_gpu_float = cupy.asarray(dirty_img_float)
        psf_gpu_float = cupy.asarray(psf_float)
        # cbeam_details_gpu_float = cupy.asarray(cbeam_details_float)
        clean_comp_gpu_float = cupy.zeros_like(dirty_img_gpu_float)
        residual_gpu_float = cupy.zeros_like(dirty_img_gpu_float)
        skymodel_gpu_float = cupy.zeros_like(dirty_img_gpu_float)

        hogbom_clean(
            dirty_img_gpu_float,
            psf_gpu_float,
            cbeam_details_float,
            loop_gain,
            threshold,
            cycle_limit,
            clean_comp_gpu_float,
            residual_gpu_float,
            skymodel_gpu_float,
        )

        clean_model_check_float = cupy.asnumpy(clean_comp_gpu_float)
        residual_check_float = cupy.asnumpy(residual_gpu_float)
        skymodel_check_float = cupy.asnumpy(skymodel_gpu_float)

        np.testing.assert_array_almost_equal(
            clean_model_check_float, clean_model_reference, decimal=4
        )
        np.testing.assert_array_almost_equal(
            residual_check_float, residual_reference, decimal=4
        )
        np.testing.assert_array_almost_equal(
            skymodel_check_float, skymodel_reference, decimal=4
        )

        print("Hogbom CLEAN float precision on GPU: Test passed")

File: tests/clean/test_hogbom_clean.cpp
/* See the LICENSE file at the top-level directory of this distribution. */

#ifdef NDEBUG
#undef NDEBUG
#endif

#include <cmath>
#include <complex>
#include <cassert>

#include "ska-sdp-func/clean/sdp_hogbom_clean.h"
#include "ska-sdp-func/utility/sdp_device_wrapper.h"
#include "ska-sdp-func/utility/sdp_logging.h"
#include "ska-sdp-func/utility/sdp_mem.h"
#include "ska-sdp-func/grid_data/sdp_gridder_uvw_es_fft.h"
#include "ska-sdp-func/visibility/sdp_dft.h"

using std::complex;

#define INDEX_2D(N2, N1, I2, I1) (N1 * I2 + I1)
#define INDEX_3D(N3, N2, N1, I3, I2, I1) (N1 * (N2 * I3 + I2) + I1)
#define INDEX_4D(N4, N3, N2, N1, I4, I3, I2, I1) \
    (N1 * (N2 * (N3 * I4 + I3) + I2) + I1)

static void create_test_data(
    sdp_Mem *dirty_img,
    sdp_Mem *psf,
    int nxydirty,
    int nxypsf,
    sdp_Error *status)
{

    const int num_components = 10;
    const int num_pols = 1;
    const int num_channels = 1;
    const int num_baselines = 1000;
    const int num_times = 1;
    const double channel_start_hz = 100e6;
    const double channel_step_hz = 100e3;

    // // calculate useful shapes and sizes
    // const int64_t dirty_img_size = dirty_img_dim * dirty_img_dim;
    // const int64_t psf_size = psf_dim * psf_dim;
    // const int64_t dirty_img_shape[] = {dirty_img_dim, dirty_img_dim};
    // const int64_t psf_shape[] = {psf_dim, psf_dim};
    const int64_t vis_shape[] = {num_times, num_baselines, num_channels, num_pols};
    const int64_t fluxes_shape[] = {num_components, num_channels, num_pols};
    const int64_t fluxes_psf_shape[] = {1, num_channels, num_pols};
    const int64_t directions_shape[] = {num_components, 3};
    const int64_t directions_psf_shape[] = {1, 3};
    const int64_t uvw_shape[] = {num_times, num_baselines, 3};

    // initialise empty arrays for dirty image generation
    sdp_Mem *vis = sdp_mem_create(SDP_MEM_COMPLEX_DOUBLE, SDP_MEM_CPU, 4, vis_shape, status);
    complex<double> *vis_ptr = (complex<double> *)sdp_mem_data(vis);
    sdp_mem_clear_contents(vis, status);

    sdp_Mem *vis_psf = sdp_mem_create(SDP_MEM_COMPLEX_DOUBLE, SDP_MEM_CPU, 4, vis_shape, status);
    complex<double> *vis_psf_ptr = (complex<double> *)sdp_mem_data(vis_psf);
    sdp_mem_clear_contents(vis_psf, status);

    sdp_Mem *fluxes = sdp_mem_create(SDP_MEM_COMPLEX_DOUBLE, SDP_MEM_CPU, 3, fluxes_shape, status);
    complex<double> *fluxes_ptr = (complex<double> *)sdp_mem_data(fluxes);
    sdp_mem_clear_contents(fluxes, status);

    sdp_Mem *fluxes_psf = sdp_mem_create(SDP_MEM_COMPLEX_DOUBLE, SDP_MEM_CPU, 3, fluxes_psf_shape, status);
    complex<double> *fluxes_psf_ptr = (complex<double> *)sdp_mem_data(fluxes_psf);
    sdp_mem_clear_contents(fluxes_psf, status);
    // Create flux of 1 + 0j's for psf
    fluxes_psf_ptr[0] = std::complex<double>(1,0);

    sdp_Mem *directions = sdp_mem_create(SDP_MEM_DOUBLE, SDP_MEM_CPU, 2, directions_shape, status);
    double *directions_ptr = (double *)sdp_mem_data(directions);
    sdp_mem_clear_contents(directions, status);

    sdp_Mem *directions_psf = sdp_mem_create(SDP_MEM_DOUBLE, SDP_MEM_CPU, 2, directions_psf_shape, status);
    // double *directions_psf_ptr = (double *)sdp_mem_data(directions_psf);
    sdp_mem_clear_contents(directions_psf, status);

    sdp_Mem *uvw = sdp_mem_create(SDP_MEM_DOUBLE, SDP_MEM_CPU, 3, uvw_shape, status);
    double *uvw_ptr = (double *)sdp_mem_data(uvw);
    sdp_mem_clear_contents(uvw, status);

    // UVW coverage as a random filled circle of given max radius, centered on (0,0)
    const int y0 = 0, x0 = 0, r_max = 3000;
    double theta = 0, radius = 0, x = 0, y = 0;

    for (int i = 0; i < num_baselines; i++)
    {

        // random angle
        theta = 2 * M_PI * (rand() / ((double)RAND_MAX));

        // random radius
        radius = r_max * (rand() / ((double)RAND_MAX));

        x = x0 + radius * cos(theta);
        y = y0 + radius * sin(theta);

        const unsigned int i_uvw = INDEX_3D(num_times, num_baselines, 3, 0, i, 0);
        uvw_ptr[i_uvw] = x;
        uvw_ptr[i_uvw + 1] = y;
        uvw_ptr[i_uvw + 2] = 0;
    }

    // Create random fluxes between 1 + 0j and 10 + 0j
    int upper = 10, lower = 1;
    double ran_flux = 0;

    for (int i = 0; i < num_components; i++)
    {

        ran_flux = (rand() % (upper - lower + 1)) + lower;

        const unsigned int i_fluxes = INDEX_3D(num_components, num_channels, num_pols, i, 0, 0);
        fluxes_ptr[i_fluxes] = complex<double>(ran_flux, 0);
    }

    // create random lmn co-ordinates between (-0.015,-0.015) and (0.015,0.015)
    upper = 0, upper = 30;
    double ran_l = 0, ran_m = 0, ran_n = 0;
    for (int i = 0; i < num_components; i++)
    {

        ran_l = (double)((rand() % (upper - lower + 1)) + lower - 15) / 100;
        ran_m = (double)((rand() % (upper - lower + 1)) + lower - 15) / 100;
        ran_n = 1 - pow(ran_l, 2) - pow(ran_m, 2);

        const unsigned int i_directions = INDEX_2D(num_components, 3, i, 0);
        directions_ptr[i_directions] = ran_l;
        directions_ptr[i_directions + 1] = ran_m;
        directions_ptr[i_directions + 2] = ran_n;
    }

    // create visabilities
    sdp_dft_point_v01(directions, fluxes, uvw, channel_start_hz, channel_step_hz, vis, status);

    // create PSF visabilities
    sdp_dft_point_v01(directions_psf, fluxes_psf, uvw, channel_start_hz, channel_step_hz, vis_psf, status);

    // initialise settings for gridder
    int fov = 2; // degrees
    double pixel_size_rad = fov * M_PI / 180 / nxydirty;
    double pixel_size_rad_psf = fov * M_PI / 180 / nxypsf;
    bool do_w_stacking = false;
    double epsilon = 1e-5;
    int num_vis = num_baselines * num_channels;

    const int64_t freqs_shape[] = {num_channels};
    sdp_Mem *freqs = sdp_mem_create(SDP_MEM_DOUBLE, SDP_MEM_CPU, 1, freqs_shape, status);
    double *freqs_ptr = (double *)sdp_mem_data(freqs);

    for (int i = 0; i < num_channels; i++)
    {
        freqs_ptr[i] = channel_start_hz + (i * channel_step_hz);
    }

    // DFT:
    // uvw is 3D and real-valued, with shape: [ num_times, num_baselines, 3 ]
    // vis is 4D and complex-valued, with shape: [ num_times, num_baselines, num_channels, num_pols ]
    // Gridder:
    // uvw [in] The (u,v,w) coordinates.  Must be complex with shape [num_rows, 3]
    // vis [in, out] The visibility data.  Must be complex with shape [num_rows, num_chan]

    // reshape vis output from DFT to fit input of gridder
    const int64_t vis_grid_shape[] = {num_baselines, num_channels}; // num_times = 1 so can be ignored, num_rows = num_baselines
    sdp_Mem *vis_grid = sdp_mem_create(SDP_MEM_COMPLEX_DOUBLE, SDP_MEM_CPU, 2, vis_grid_shape, status);
    complex<double> *vis_grid_ptr = (complex<double> *)sdp_mem_data(vis_grid);

    for (int i = 0; i < num_baselines; i++)
    {

        const unsigned int i_vis = INDEX_4D(num_times, num_baselines, num_channels, num_pols, 0, i, 0, 0);
        const unsigned int i_vis_grid = INDEX_2D(num_baselines, num_channels, i, 0);
        vis_grid_ptr[i_vis_grid] = vis_ptr[i_vis];
    }

    sdp_Mem *vis_grid_psf = sdp_mem_create(SDP_MEM_COMPLEX_DOUBLE, SDP_MEM_CPU, 2, vis_grid_shape, status);
    complex<double> *vis_grid_psf_ptr = (complex<double> *)sdp_mem_data(vis_grid_psf);

    for (int i = 0; i < 1; i++)
    {

        const unsigned int i_vis = INDEX_4D(num_times, num_baselines, num_channels, num_pols, 0, i, 0, 0);
        const unsigned int i_vis_grid = INDEX_2D(num_baselines, num_channels, i, 0);
        vis_grid_psf_ptr[i_vis_grid] = vis_psf_ptr[i_vis];
    }

    // reshape UVW to fit input of gridder
    const int64_t uvw_grid_shape[] = {num_baselines, 3}; // num_times = 1 so can be ignored, num_rows = num_baselines
    sdp_Mem *uvw_grid = sdp_mem_create(SDP_MEM_DOUBLE, SDP_MEM_CPU, 2, uvw_grid_shape, status);
    double *uvw_grid_ptr = (double *)sdp_mem_data(uvw_grid);

    for (int i = 0; i < num_baselines; i++)
    {

        const unsigned int i_uvw = INDEX_3D(num_times, num_baselines, 3, 0, i, 0);
        const unsigned int i_uvw_grid = INDEX_2D(num_baselines, 3, i, 0);
        uvw_ptr[i_uvw] = uvw_grid_ptr[i_uvw_grid];
        uvw_ptr[i_uvw + 1] = uvw_grid_ptr[i_uvw_grid + 1];
        uvw_ptr[i_uvw + 2] = uvw_grid_ptr[i_uvw_grid + 2];
    }

    // copy to gpu becaause gridder only works on gpu
    sdp_Mem *freqs_gpu = sdp_mem_create_copy(freqs, SDP_MEM_GPU, status);
    sdp_Mem *uvw_gpu = sdp_mem_create_copy(uvw_grid, SDP_MEM_GPU, status);
    sdp_Mem *vis_gpu = sdp_mem_create_copy(vis_grid, SDP_MEM_GPU, status);
    sdp_Mem *vis_psf_gpu = sdp_mem_create_copy(vis_grid_psf, SDP_MEM_GPU, status);
    sdp_Mem *dirty_img_gpu = sdp_mem_create_copy(dirty_img, SDP_MEM_GPU, status);
    sdp_Mem *psf_gpu = sdp_mem_create_copy(psf, SDP_MEM_GPU, status);


    const int64_t weight_shape[] = {num_vis, num_channels};
    sdp_Mem *weight = sdp_mem_create(SDP_MEM_DOUBLE, SDP_MEM_CPU, 2, weight_shape, status);
    double *weight_ptr = (double *)sdp_mem_data(weight);

    for (int i = 0; i < (num_vis * num_channels); i++)
    {
        weight_ptr[i] = 1;
    }

    sdp_Mem *weight_gpu = sdp_mem_create_copy(weight, SDP_MEM_GPU, status);

    // create gridder for dirty image
    sdp_GridderUvwEsFft *gridder = sdp_gridder_uvw_es_fft_create_plan(
        uvw_gpu,
        freqs_gpu,
        vis_gpu,
        weight_gpu,
        dirty_img_gpu,
        pixel_size_rad,
        pixel_size_rad,
        epsilon,
        pixel_size_rad,
        pixel_size_rad,
        do_w_stacking,
        status);

    sdp_grid_uvw_es_fft(
        gridder,
        uvw_gpu,
        freqs_gpu,
        vis_gpu,
        weight_gpu,
        dirty_img_gpu,
        status);

    sdp_gridder_uvw_es_fft_free_plan(gridder);

    // create gridder for psf
    sdp_GridderUvwEsFft *gridder2 = sdp_gridder_uvw_es_fft_create_plan(
        uvw_gpu,
        freqs_gpu,
        vis_psf_gpu,
        weight_gpu,
        psf_gpu,
        pixel_size_rad_psf,
        pixel_size_rad_psf,
        epsilon,
        pixel_size_rad_psf,
        pixel_size_rad_psf,
        do_w_stacking,
        status);

    sdp_grid_uvw_es_fft(
        gridder2,
        uvw_gpu,
        freqs_gpu,
        vis_psf_gpu,
        weight_gpu,
        psf_gpu,
        status);

    sdp_gridder_uvw_es_fft_free_plan(gridder2);

    // normalise fluxes by number of baselines
    int64_t num_elements_dirty_img = sdp_mem_num_elements(dirty_img_gpu);

    sdp_mem_copy_contents(dirty_img, dirty_img_gpu, 0, 0, num_elements_dirty_img, status);
    double *dirty_img_ptr = (double *)sdp_mem_data(dirty_img);

    for (int i = 0; i < (nxydirty * nxydirty); i++)
    {
        dirty_img_ptr[i] /= num_baselines;
    }

    int64_t num_elements_psf = sdp_mem_num_elements(psf_gpu);

    sdp_mem_copy_contents(psf, psf_gpu, 0, 0, num_elements_psf, status);
    double *psf_ptr = (double *)sdp_mem_data(psf);

    for (int i = 0; i < (nxypsf * nxypsf); i++)
    {
        psf_ptr[i] /= num_baselines;
    }

    sdp_mem_ref_dec(vis);
    sdp_mem_ref_dec(vis_psf);
    sdp_mem_ref_dec(fluxes);
    sdp_mem_ref_dec(fluxes_psf);
    sdp_mem_ref_dec(directions);
    sdp_mem_ref_dec(directions_psf);
    sdp_mem_ref_dec(uvw);
    sdp_mem_ref_dec(freqs);
    sdp_mem_ref_dec(vis_grid);
    sdp_mem_ref_dec(vis_grid_psf);
    sdp_mem_ref_dec(uvw_grid);
    sdp_mem_ref_dec(freqs_gpu);
    sdp_mem_ref_dec(uvw_gpu);
    sdp_mem_ref_dec(vis_gpu);
    sdp_mem_ref_dec(vis_psf_gpu);
    sdp_mem_ref_dec(dirty_img_gpu);
    sdp_mem_ref_dec(psf_gpu);
    sdp_mem_ref_dec(weight);
    sdp_mem_ref_dec(weight_gpu);
}

static void run_and_check(
    const char *test_name,
    bool read_only_output,
    sdp_MemType input_type,
    sdp_MemType output_type,
    sdp_MemLocation input_location,
    sdp_MemLocation output_location,
    sdp_Error *status)
{

    // settings
    int nxydirty = 1024;
    int nxypsf = 2048;
    double loop_gain = 0.1;
    double threshold = 0.001;
    int cycle_limit = 10000;

    // create test data
    const int64_t diryt_img_shape[] = {nxydirty, nxydirty};
    sdp_Mem *dirty_img = sdp_mem_create(input_type, SDP_MEM_CPU, 2, diryt_img_shape, status);
    sdp_mem_clear_contents(dirty_img, status);

    const int64_t psf_shape[] = {nxypsf, nxypsf};
    sdp_Mem *psf = sdp_mem_create(input_type, SDP_MEM_CPU, 2, psf_shape, status);
    sdp_mem_clear_contents(psf, status);

    const int64_t cbeam_details_shape[] = {3};
    sdp_Mem *cbeam_details = sdp_mem_create(input_type, SDP_MEM_CPU, 1, cbeam_details_shape, status);
    double *cbeam_details_ptr = (double *)sdp_mem_data(cbeam_details);
    // pre-computed variables
    cbeam_details_ptr[0] = 23.9;
    cbeam_details_ptr[1] = 24.6;
    cbeam_details_ptr[2] = 23.6;
    
    create_test_data(
        dirty_img,
        psf,
        nxydirty,
        nxypsf,
        status);

    // create output
    const int64_t skymodel_shape[] = {nxydirty, nxydirty};
    sdp_Mem *skymodel = sdp_mem_create(output_type, SDP_MEM_CPU, 2, skymodel_shape, status);
    sdp_mem_clear_contents(skymodel, status);

    // Copy inputs to specified location.
    sdp_Mem* dirty_img_copy = sdp_mem_create_copy(dirty_img, input_location, status);
    sdp_Mem* psf_copy = sdp_mem_create_copy(psf, input_location, status);
    sdp_Mem* cbeam_details_copy = sdp_mem_create_copy(cbeam_details, input_location, status);
    sdp_Mem* skymodel_copy = sdp_mem_create_copy(skymodel, output_location, status);
    sdp_mem_set_read_only(skymodel_copy, read_only_output);


    // call function to test
    SDP_LOG_INFO("Running test: %s", test_name);
    sdp_hogbom_clean(
        dirty_img_copy,
        psf_copy,
        cbeam_details_copy,
        loop_gain,
        threshold,
        cycle_limit,
        skymodel_copy,
        status);

    sdp_mem_ref_dec(dirty_img);
    sdp_mem_ref_dec(psf);
    sdp_mem_ref_dec(cbeam_details);
    sdp_mem_ref_dec(skymodel);
    sdp_mem_ref_dec(dirty_img_copy);
    sdp_mem_ref_dec(psf_copy);
}

int main()
{
    SDP_LOG_INFO("start of test:");

    // happy paths
    {
        sdp_Error status = SDP_SUCCESS;
        run_and_check("CPU, double precision", false,
                      SDP_MEM_DOUBLE, SDP_MEM_DOUBLE,
                      SDP_MEM_CPU, SDP_MEM_CPU, &status);
        assert(status == SDP_SUCCESS);
    }

    // unhappy paths
    {
        sdp_Error status = SDP_SUCCESS;
        run_and_check("CPU, Type Mismatch", false,
                      SDP_MEM_DOUBLE, SDP_MEM_FLOAT,
                      SDP_MEM_CPU, SDP_MEM_CPU, &status);
        assert(status == SDP_ERR_DATA_TYPE);
    }

    {
        sdp_Error status = SDP_SUCCESS;
        run_and_check("CPU, Output set to read-only", true,
                      SDP_MEM_DOUBLE, SDP_MEM_DOUBLE,
                      SDP_MEM_CPU, SDP_MEM_CPU, &status);
        assert(status == SDP_ERR_RUNTIME);
    }

    {
        sdp_Error status = SDP_SUCCESS;
        run_and_check("CPU, location mis-match", false,
                      SDP_MEM_DOUBLE, SDP_MEM_DOUBLE,
                      SDP_MEM_CPU, SDP_MEM_GPU, &status);
        assert(status == SDP_ERR_MEM_LOCATION);
    }

        
    #ifdef SDP_HAVE_CUDA
    // Happy paths
    {
        sdp_Error status = SDP_SUCCESS;
        run_and_check("GPU, double precision", false,
                      SDP_MEM_DOUBLE, SDP_MEM_DOUBLE,
                      SDP_MEM_GPU, SDP_MEM_GPU, &status);
        assert(status == SDP_SUCCESS);
    }

    #endif

    return 0;
}